cpmpy/solvers/choco.py

```diff
@@ -0,0 +1,617 @@
+                                                  index)
+                elif lhs.name == 'mod':
+                    # Choco needs dividend to be a variable
+                    dividend = self.to_var(lhs.args[0])
+                    if dividend is None:
+                        raise ChocoTypeException(
+                            f"Choco does not accept {lhs.args[0]} with type {type(lhs.args[0])} in lhs of expression {lhs.name}")
+                    # Choco needs divisor to be a variable
+                    divisor = self.to_var(lhs.args[1])
+                    if divisor is None:
+                        raise ChocoTypeException(
+                            f"Choco does not accept {lhs.args[1]} with type {type(lhs.args[1])} in lhs of expression {lhs.name}")
+                    # Choco needs result to be a variable
+                    result = self.to_var(rhs)
+                    if result is None:
+                        raise ChocoTypeException(
+                            f"Choco does not accept {rhs} with type {type(rhs)} as rhs of expression {lhs.name}")
+                    return self.chc_model.mod(dividend, divisor, result)
+                elif lhs.name == 'pow':
+                    chcrhs = self.to_var(rhs)
+                    if chcrhs is None:
+                        raise ChocoTypeException(f"Choco does not accept {rhs} with type {type(rhs)} as rhs of expression {lhs.name}")
+                    return self.chc_model.pow(self.solver_vars(lhs.args[0]), self.solver_vars(lhs.args[1]),
+                                              chcrhs)
+            raise NotImplementedError(
+                "Not a known supported Choco left-hand-side '{}' {}".format(lhs.name, cpm_expr))
+
+        # base (Boolean) global constraints
+        elif isinstance(cpm_expr, GlobalConstraint):
+
+            if cpm_expr.name == 'alldifferent':
+                vars = self.to_vars(cpm_expr.args)
+                if any(c is None for c in vars):
+                    raise ChocoTypeException(f"Choco cannot accept alldifferent with: {vars}")
+                return self.chc_model.all_different(vars)
+            elif cpm_expr.name == 'alldifferent_except0':
+                vars = self.to_vars(cpm_expr.args)
+                if any(c is None for c in vars):
+                    raise ChocoTypeException(f"Choco cannot accept alldifferent_except0 with: {vars}")
+                return self.chc_model.all_different_except_0(vars)
+            elif cpm_expr.name == 'allequal':
+                vars = self.to_vars(cpm_expr.args)
+                if any(c is None for c in vars):
+                    raise ChocoTypeException(f"Choco cannot accept allequal with: {vars}")
+                return self.chc_model.all_equal(vars)
+            elif cpm_expr.name == 'table':
+                assert (len(cpm_expr.args) == 2)  # args = [array, table]
+                array, table = self.solver_vars(cpm_expr.args)
```

Comment:
This is also gonne have to become to_vars in stead of solver_vars for the array component. Not sure about the table component though..

Reply:
maybe not logical but we allow something like Table([var,5],[(2,5),(3,6)])

Reply:
tbh I am not sure we should. first args should be vars always, second one the allowed tuples

Reply:
Choco table constraint only accepts list of list of ints as second argument, so we don't want to use to_var as it will make it a list of list of vars

Reply:
I think you are right, will make a pr to put a check in the constructor of table

Reply:
pr (https://github.com/CPMpy/cpmpy/pull/428) ;)